### PR TITLE
Undo method that does not save changes to object;

### PR DESCRIPTION
**Pull request by alexkravets**
This make it possible to get specific version of an object: ```object.undo(nill, from: <version>, to: object.version)```

### CODE REVIEW

**Comment by dblock** [comment]
This needs CHANGELOG and README updates, please.

File: lib/mongoid/history/trackable.rb [hunk 1]
```diff
@@ -71,6 +71,16 @@ def history_tracks
...
+        def undo(modifier = nil, options_or_version = nil)
+          versions = get_versions_criteria(options_or_version).to_a
+          versions.sort! { |v1, v2| v2.version <=> v1.version }
+
+          versions.each do |v|
+            undo_attr       = v.undo_attr(modifier)
+            self.attributes = undo_attr
+          end
+        end
+
```
Comment (dblock) [inline lib/mongoid/history/trackable.rb]: It's extremely similar to `undo!`. I think you need to rewrite `undo!` as `undo` and `save!`.

**Comment by alexkravets** [comment]
Thank you for the heads up! Tests, CHANGELOG, and README updates added, please take a look.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 0.4.5 (Next)
 ------------
 
+* Add `undo` method, that helps to get specific version of an object without saving changes - [@alexkravets](https://github.com/alexkravets).
```
Comment (dblock) [inline CHANGELOG.md]: Amend with the PR link please, like below.

File: README.md [hunk 3]
```diff
@@ -121,6 +121,9 @@ track.undo! user # comment title should be "Test"
 
 track.redo! user # comment title should be "Test 2"
 
+# undo comment to version 1 without save
+comment.undo nill, from: 1, to: comment.version
```
Comment (dblock) [inline README.md]: What is `nill` with two ll's :)?

File: lib/mongoid/history/trackable.rb [hunk 4]
```diff
@@ -74,21 +74,41 @@ def history_tracks
...
 
           versions.each do |v|
-            undo_attr = v.undo_attr(modifier)
-            if Mongoid::History.mongoid3? # update_attributes! not bypassing rails 3 protected attributes
-              assign_attributes(undo_attr, without_protection: true)
-              save!
-            else # assign_attributes with 'without_protection' option does not work with rails 4/mongoid 4
-              update_attributes!(undo_attr)
-            end
+            # NOTE: not sure how this works with different versions on Mongoid,
```
Comment (dblock) [inline lib/mongoid/history/trackable.rb]: Can you elaborate on the note please?

File: lib/mongoid/history/trackable.rb [hunk 5]
```diff
@@ -74,21 +74,41 @@ def history_tracks
...
+            #       works for rails 4/ mongoid 4
+            undo_attr       = v.undo_attr(modifier)
+            self.attributes = undo_attr
```
Comment (dblock) [inline lib/mongoid/history/trackable.rb]: I think you don't need to assign `undo_attr`, just do `self.attributes = v....`.

File: lib/mongoid/history/trackable.rb [hunk 6]
```diff
@@ -74,21 +74,41 @@ def history_tracks
...
 
+        #  undo! :from => 1, :to => 5
+        #  undo! 4
+        #  undo! :last => 10
+        def undo!(modifier = nil, options_or_version = nil)
+          undo(modifier, options_or_version)
+          save!
+        end
+
+        # def undo!(modifier = nil, options_or_version = nil)
```
Comment (dblock) [inline lib/mongoid/history/trackable.rb]: Delete commented code.

File: spec/integration/integration_spec.rb [hunk 7]
```diff
@@ -627,6 +627,60 @@ class Foo < Comment
       end
 
       describe 'undo' do
+        [nil].each do |method|
+          context "#{method || 'instance'}" do
+            it 'should recognize :from, :to options' do
```
Comment (dblock) [inline spec/integration/integration_spec.rb]: Generally change "it should recognize ..." to "it recognizes ..."

File: spec/integration/integration_spec.rb [hunk 8]
```diff
@@ -627,6 +627,60 @@ class Foo < Comment
       end
 
       describe 'undo' do
+        [nil].each do |method|
```
Comment (dblock) [inline spec/integration/integration_spec.rb]: Collapse tests for `undo` and `undo!` together, maybe with two loops, one over method/instance and inside over `undo!` and `undo!`. I also would like an explicit test that `undo!` is just `undo` + `save!`.

**Comment by dblock** [comment]
See my comments, it's getting closer. Make sure to fix the build and once everything is cleaned up, squash your commits, please.

**Comment by alexkravets** [comment]
"I also would like an explicit test that undo! is just undo + save!." — could you please provide an example of test you want to have for it? Thanks.

**Comment by dblock** [comment]
This is good, squash your commits and this is good to go.

The undo/save test would still be nice. I am thinking something like this:

```ruby
it "implements undo! in terms of undo and save!" do
    commment = ...
    expect(comment).to receive(:undo)
    expect(comment).to receive(:save!)
    comment.undo! ...
end
```

**Comment by dblock** [comment]
Merged, thank you.

**Pull request merged**